### ISSUE TITLE
Drop Qt5 support completely for 2.5 & main

### ISSUE THREAD

**Issue by cr7pt0gr4ph7** [mixxxdj/mixxx#13826]
As it [came up](https://github.com/mixxxdj/mixxx/pull/13807#issuecomment-2453000013) in a PR discussion, it seems that the issue of whether to support both Qt5+6, or only Qt6, has not been entirely decided yet.

- While versions up to and including `2.4` are built with Qt5,
- the upcoming `2.5` version as well as the current `main` are built with Qt6 instead.

To clarify, this proposal only applies to `main`, and not to the older branches.

In my opinion, there is no sense in keeping the remnants of Qt5 support around if we are not actively making sure that Mixxx can actually still be built (and run) with Qt5 using appropriate CI jobs. Otherwise, the code base will silently develop into a state where it either will not build on Qt5 anymore, or (IMO worse), will have buggy code that is only active when building & running on Qt6.

So, with the aim of cleaner code, less bugs and improved readability in mind, I propose we decide for one of three options:
1. **Keep Qt5 support around.** Some work would still need to be done:
  - Clarify the coding guidelines to reflect that
  - Make sure that PRs are reviewed for Qt5 support
  - Establish a Qt5 build job that checks for compatibility & runs unit tests.
2. **Drop Qt6 support in a concerted effort.** So, create a big PR that removes all `#if QT_VERSION ...` code from all files.
  - Pro: Everything is clean after that, so new people reading the code won't be confused.
  - Con: I anticipate a lot of merge conflicts related to long-running PRs.
  - Con: I think it's easier to mess up but harder to review and test if this is done in one big PR.
3. **Gradually remove Qt6 support if you're touching a file anyway.** (which is what I would prefer).
  - Pro: Merge conflicts are limited to files that are would probably have conflicts anyway.
  - Pro: If you're touching some functionality, you are hopefully testing it anyway.
  - Pro: Reviews are easier if you're not changing too much code at once.
  - Con: No big, clean cut - people might need to read the documentation to see our guidance on Qt versions.
  - We can also do focused cleanups for certain areas of functionality in separate PRs if desired, to speed up the removal process.

As an additional datapoint, a [quick GitHub search](https://github.com/search?q=repo%3Amixxxdj%2Fmixxx+%23if+QT_VERSION&type=code) uncovered 77 files with some variation of `#if QT_VERSION`, some with multiple occurences per file (the most I found was 19 occurences in a single file).

Questions:
- Is there anybody who wants/needs new Mixxx versions to be able to build and run on Qt5?
- Does keeping the old code around make backports easier? Although the issue with "untested" code would apply in that case.

**Comment by cr7pt0gr4ph7** [comment]
For reference, see also the [Minimum requirements policy](https://github.com/mixxxdj/mixxx/wiki/Minimum-requirements-policy) written down in the Wiki as well as [this Zulip discussion]([this Zulip discussion](https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/minimum.20requirements.20policy/near/203415058) from 2020.

**Comment by m0dB** [comment]
Note that Qt 5 officially reaches end of life in May 2025.

**Comment by daschuer** [comment]
Some time ago we have discussed that and failed to communicate the decision clearly. 

* 2.4.2 will be the last Qt5 release 
* 2.5.0 will be kept compatible with Qt5 for the sake of regression testing.
* main is already Qt6 only 

The Qt5 support in main will not likely disappear slowly, because it is unlikely to be touched. If we want to get rid of it, we need to do it explicit. This comes with extra efforts of doing it, review/testing and introduced conflicts when merging from 2.5. 

Let's join forces now, to release 2.5 instead. Once it is released we may rip of Qt6 from main entirely. 

Can we change thd issue title accordingly?

**Comment by cr7pt0gr4ph7** [comment]
> Some time ago we have discussed that and failed to communicate the decision clearly.
> 
> * 2.4.2 will be the last Qt5 release
> * 2.5.0 will be kept compatible with Qt5 for the sake of regression testing.
> * main is already Qt6 only

Communication in big projects is always a challenge, especially with people coming and going. I apologize in advance for asking questions that have already been answered. But at least it's a good way to discover what needs to be documented (at all/explicitly/more clearly/in a more visible place/...) :+1: 

> The Qt5 support in main will not likely disappear slowly, because it is unlikely to be touched. If we want to get rid of it, we need to do it explicit. This comes with extra efforts of doing it, review/testing and introduced conflicts when merging from 2.5.

I kindly beg to differ with regard to your first point, because trying to do that is what ultimately led me to create this issue (see [this PR discussion](https://github.com/mixxxdj/mixxx/pull/13807#issuecomment-2453000013), [this review comment](https://github.com/mixxxdj/mixxx/pull/13816#discussion_r1825624268), some of [@ronso0]s personal branches, and some more code in my pipeline).

I'm mainly concerned with "Can I remove these `#if` directives when I'm cleaning up code prior to a bigger refactoring". Unnecessary `#if`s just mean "dead code that isn't used (and also not maintained)" aka. a readability hindrance that I primarily care about in code that I'm trying to make more readable.

I agree with the added effort for removing this code for the sake of it, which is why I would propose (as explained in the issue text):
1. Allowing removal of the `#if`s in feature & refactoring PRs (which should not be merged into `2.5` while trying to stabilize that release) that touch that code (and would cause merge conflicts) anyway.
2. Preventing addition of new `#if`s for Qt versions < 6 in such PRs.
3. For now: Disallowing removal in bugfix PRs, as per the current policy.

> Let's join forces now, to release 2.5 instead. Once it is released we may rip of Qt6 from main entirely.

I'm running a custom `main`-based build of Mixxx (inside a Flatpak) as my production setup that I periodically update to the most-recent `main` + my own changes using some custom scripts, so I'm primarily concerned with `main`, but I'll try helping out with review & testing :+1:.

> Can we change thd issue title accordingly?

Done :+1: Thanks for the feedback!

**Comment by daschuer** [comment]
@ronso0 Do you have a demand for keeping Qt5 in main? 
I think it is no longer necessary, since it is broken anyway. Or did we repair it lately.

**Comment by ronso0** [comment]
Nope, I can't even remember anymore that I built 2.5 or main with Qt5.
FWIW I didn't understand why @cr7pt0gr4ph7 referred to my personal branches. My 'production' branch is based on main with all my feature/fix branches merged commit by commit and I don't rely on Qt5.

So :+1: for removing Qt5 switches/hacks in main asap.

**Comment by cr7pt0gr4ph7** [comment]
> FWIW I didn't understand why @cr7pt0gr4ph7 referred to my personal branches.

Sorry for the confusion, I meant some instances where you removed `⁰

**Comment by ronso0** [comment]
First (and biggest) step is done in #15251 
A follow-up should take of the QMutex compat stuff  mentioned there.
